SerializationException when storing ScheduleToken in a Saga using JacksonSerializer

The ```EventScheduler``` interface returns a ```ScheduleToken``` type as a return to ```schedule()```. The problem is when a Saga is beign loaded, the token cannot be deserialized cause ```ScheduleToken``` is an abstract class.

Possible solutions are:
- enable default typing on jackson (personally i don't like this one)
- provide ```JsonTypeInfo``` on ```ScheduleToken```
- implement a mechanism to provide own serializer/deserializer for ```ScheduleToken``` (possibly using previous approach as a default)

Below is a stack trace of an error.
```
org.axonframework.serialization.SerializationException: Error while deserializing object
	at org.axonframework.serialization.json.JacksonSerializer.deserialize(JacksonSerializer.java:214) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.saga.repository.jpa.JpaSagaStore.loadSaga(JpaSagaStore.java:150) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.saga.repository.AnnotatedSagaRepository.doLoadSaga(AnnotatedSagaRepository.java:251) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.saga.repository.AnnotatedSagaRepository.lambda$doLoad$1(AnnotatedSagaRepository.java:131) ~[axon-core-3.2.1.jar:3.2.1]
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660) ~[na:1.8.0_144]
	at org.axonframework.eventhandling.saga.repository.AnnotatedSagaRepository.doLoad(AnnotatedSagaRepository.java:130) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.saga.repository.AnnotatedSagaRepository.doLoad(AnnotatedSagaRepository.java:46) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.saga.repository.LockingSagaRepository.load(LockingSagaRepository.java:54) ~[axon-core-3.2.1.jar:3.2.1]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[na:1.8.0_144]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[na:1.8.0_144]
	at java.util.TreeMap$KeySpliterator.forEachRemaining(TreeMap.java:2746) ~[na:1.8.0_144]
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580) ~[na:1.8.0_144]
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270) ~[na:1.8.0_144]
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1548) ~[na:1.8.0_144]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[na:1.8.0_144]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[na:1.8.0_144]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[na:1.8.0_144]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_144]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[na:1.8.0_144]
	at org.axonframework.eventhandling.saga.AbstractSagaManager.handle(AbstractSagaManager.java:70) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.AbstractEventProcessor.lambda$null$2(AbstractEventProcessor.java:139) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.messaging.DefaultInterceptorChain.proceed(DefaultInterceptorChain.java:57) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.messaging.interceptors.CorrelationDataInterceptor.handle(CorrelationDataInterceptor.java:65) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.messaging.DefaultInterceptorChain.proceed(DefaultInterceptorChain.java:55) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.TrackingEventProcessor.lambda$new$1(TrackingEventProcessor.java:176) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.messaging.DefaultInterceptorChain.proceed(DefaultInterceptorChain.java:55) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.messaging.interceptors.TransactionManagingInterceptor.handle(TransactionManagingInterceptor.java:50) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.messaging.DefaultInterceptorChain.proceed(DefaultInterceptorChain.java:55) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.AbstractEventProcessor.lambda$processInUnitOfWork$3(AbstractEventProcessor.java:146) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.messaging.unitofwork.BatchingUnitOfWork.executeWithResult(BatchingUnitOfWork.java:80) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.AbstractEventProcessor.processInUnitOfWork(AbstractEventProcessor.java:135) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.TrackingEventProcessor.processBatch(TrackingEventProcessor.java:282) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.TrackingEventProcessor.processingLoop(TrackingEventProcessor.java:209) ~[axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.TrackingEventProcessor$TrackingSegmentWorker.run(TrackingEventProcessor.java:620) [axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.TrackingEventProcessor$WorkerLauncher.run(TrackingEventProcessor.java:713) [axon-core-3.2.1.jar:3.2.1]
	at org.axonframework.eventhandling.TrackingEventProcessor$CountingRunnable.run(TrackingEventProcessor.java:547) [axon-core-3.2.1.jar:3.2.1]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_144]
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.axonframework.eventhandling.scheduling.ScheduleToken` (no Creators, like default construct, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
 at [Source: (byte[])"..."; line: 1, column: 176] (through reference chain: custis.modeus.notification.component.notification.command.saga.PersonNotificationPublishingSaga["retryToken"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1451) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1027) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserialize(AbstractDeserializer.java:265) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:136) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:287) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1611) ~[jackson-databind-2.9.3.jar:2.9.3]
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1234) ~[jackson-databind-2.9.3.jar:2.9.3]
	at org.axonframework.serialization.json.JacksonSerializer.deserialize(JacksonSerializer.java:212) ~[axon-core-3.2.1.jar:3.2.1]
```

--- a/core/src/main/java/org/axonframework/eventhandling/scheduling/ScheduleToken.java
+++ b/core/src/main/java/org/axonframework/eventhandling/scheduling/ScheduleToken.java
@@ -16,6 +16,8 @@
 
 package org.axonframework.eventhandling.scheduling;
 
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+
 import java.io.Serializable;
 
 /**
@@ -25,6 +27,7 @@ import java.io.Serializable;
  * @author Allard Buijze
  * @since 0.7
  */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
 public interface ScheduleToken extends Serializable {
 
 }

--- a/core/src/main/java/org/axonframework/eventhandling/scheduling/java/SimpleScheduleToken.java
+++ b/core/src/main/java/org/axonframework/eventhandling/scheduling/java/SimpleScheduleToken.java
@@ -16,7 +16,11 @@
 
 package org.axonframework.eventhandling.scheduling.java;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
 import org.axonframework.eventhandling.scheduling.ScheduleToken;
+
+import java.util.Objects;
 
 /**
  * ScheduleToken for tasks event scheduled using the SimpleEventScheduler.
@@ -34,7 +38,8 @@ public class SimpleScheduleToken implements ScheduleToken {
      *
      * @param tokenId The identifier referencing the scheduled task.
      */
-    public SimpleScheduleToken(String tokenId) {
+    @JsonCreator
+    public SimpleScheduleToken(@JsonProperty("tokenId") String tokenId) {
         this.tokenId = tokenId;
     }
 
@@ -45,5 +50,29 @@ public class SimpleScheduleToken implements ScheduleToken {
      */
     public String getTokenId() {
         return tokenId;
+    }
+
+    @Override
+    public String toString() {
+        return "SimpleScheduleToken{" +
+                "tokenId='" + tokenId + '\'' +
+                '}';
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tokenId);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        final SimpleScheduleToken other = (SimpleScheduleToken) obj;
+        return Objects.equals(this.tokenId, other.tokenId);
     }
 }

--- a/core/src/main/java/org/axonframework/eventhandling/scheduling/quartz/QuartzScheduleToken.java
+++ b/core/src/main/java/org/axonframework/eventhandling/scheduling/quartz/QuartzScheduleToken.java
@@ -16,7 +16,11 @@
 
 package org.axonframework.eventhandling.scheduling.quartz;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
 import org.axonframework.eventhandling.scheduling.ScheduleToken;
+
+import java.util.Objects;
 
 import static java.lang.String.format;
 
@@ -39,7 +43,9 @@ public class QuartzScheduleToken implements ScheduleToken {
      * @param jobIdentifier   The identifier used when registering the job with quartz.
      * @param groupIdentifier The identifier of the group the job is part of.
      */
-    public QuartzScheduleToken(String jobIdentifier, String groupIdentifier) {
+    @JsonCreator
+    public QuartzScheduleToken(@JsonProperty("jobIdentifier") String jobIdentifier,
+                               @JsonProperty("groupIdentifier") String groupIdentifier) {
         this.jobIdentifier = jobIdentifier;
         this.groupIdentifier = groupIdentifier;
     }
@@ -65,5 +71,23 @@ public class QuartzScheduleToken implements ScheduleToken {
     @Override
     public String toString() {
         return format("Quartz Schedule token for job [%s] in group [%s]", jobIdentifier, groupIdentifier);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(jobIdentifier, groupIdentifier);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        final QuartzScheduleToken other = (QuartzScheduleToken) obj;
+        return Objects.equals(this.jobIdentifier, other.jobIdentifier)
+                && Objects.equals(this.groupIdentifier, other.groupIdentifier);
     }
 }

--- a/core/src/test/java/org/axonframework/eventhandling/scheduling/ScheduleTokenJacksonSerializationTest.java
+++ b/core/src/test/java/org/axonframework/eventhandling/scheduling/ScheduleTokenJacksonSerializationTest.java
@@ -1,0 +1,37 @@
+package org.axonframework.eventhandling.scheduling;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.axonframework.eventhandling.scheduling.java.SimpleScheduleToken;
+import org.axonframework.eventhandling.scheduling.quartz.QuartzScheduleToken;
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+
+public class ScheduleTokenJacksonSerializationTest {
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Test
+    public void testSerializationAndDeserializationOfSimpleScheduleToken() throws IOException {
+        ScheduleToken scheduleToken = new SimpleScheduleToken("tokenId");
+
+        String serializationResult = objectMapper.writeValueAsString(scheduleToken);
+        SimpleScheduleToken deserializationResult = objectMapper.readerFor(SimpleScheduleToken.class)
+                                                                .readValue(serializationResult);
+
+        assertEquals(scheduleToken, deserializationResult);
+    }
+
+    @Test
+    public void testSerializationAndDeserializationOfQuartzScheduleToken() throws IOException {
+        ScheduleToken scheduleToken = new QuartzScheduleToken("jobIdentifier", "groupIdentifier");
+
+        String serializationResult = objectMapper.writeValueAsString(scheduleToken);
+        QuartzScheduleToken deserializationResult = objectMapper.readerFor(QuartzScheduleToken.class)
+                                                                .readValue(serializationResult);
+
+        assertEquals(scheduleToken, deserializationResult);
+    }
+}

